test/test_quantized.py

```diff
@@ -311,6 +311,113 @@ def test_cat(self, X, num, axis, relu):
             cat_q = q_cat_op(tensors_q, axis=axis, scale=scale,
                              zero_point=zero_point)
 
+
+@unittest.skipIf(
+    TEST_WITH_UBSAN or not torch.fbgemm_is_cpu_supported(),
+    " Quantized Linear requires FBGEMM. FBGEMM does not play"
+    " well with UBSAN at the moment, so we skip the test if"
+    " we are in a UBSAN environment.",
+)
+class TestDynamicQuantizedLinear(TestCase):
+    """Tests the correctness of the dynamic quantized linear and linear_relu op."""
+    @given(
+        batch_size=st.integers(1, 4),
+        input_channels=st.integers(16, 32),
+        output_channels=st.integers(4, 8),
+        use_bias=st.booleans(),
+        use_relu=st.booleans(),
+    )
+    def test_qlinear(self, batch_size, input_channels, output_channels, use_bias, use_relu):
+        qlinear_prepack = torch.ops.quantized.fbgemm_linear_prepack
+        if use_relu:
+            qlinear_dynamic = torch.ops.quantized.fbgemm_linear_relu_dynamic
+        else:
+            qlinear_dynamic = torch.ops.quantized.fbgemm_linear_dynamic
+
+        X_scale = 1.0
+        X_zp = 0
+        X_value_min = 0
+        X_value_max = 255
+        X_q0 = np.round(np.random.rand(batch_size, input_channels) * (
+            X_value_max - X_value_min) + X_value_min
+        ).astype(np.uint8)
+        X_q0[0, 0] = X_value_min
+        X_q0[0, 1] = X_value_max
+
+        W_scale = 1.0
+        W_zp = 0
+        W_value_min = -128
+        W_value_max = 127
+        W_q0 = np.round(
+            np.random.rand(output_channels, input_channels)
+            * (W_value_max - W_value_min)
+            + W_value_min
+        ).astype(np.int8)
+        W_q0[0, 0] = W_value_min
+        W_q0[1, 0] = W_value_max
+
+        b_value_min = -10
+        b_value_max = 10
+        b_q0 = np.round(
+            np.random.rand(output_channels) * (b_value_max - b_value_min) +
+            b_value_min
+        ).astype(np.int32) if use_bias else None
+
+        avoid_vpmaddubsw_overflow_linear(
+            batch_size,
+            input_channels,
+            output_channels,
+            X_q0,
+            X_value_min,
+            X_value_max,
+            W_q0,
+            W_value_min,
+            W_value_max,
+        )
+
+        X_fp32 = torch.from_numpy(_dequantize(X_q0, X_scale, X_zp)).to(dtype=torch.float)
+        W_fp32 = torch.from_numpy(_dequantize(W_q0, W_scale, W_zp)).to(dtype=torch.float)
+        b_fp32 = torch.from_numpy(
+            _dequantize(b_q0, X_scale * W_scale, 0)
+        ).to(dtype=torch.float) if use_bias else None
+
+        W_scale, W_zp = _calculate_dynamic_qparams(W_fp32, torch.qint8)
+        W_q = torch.quantize_linear(W_fp32, scale=W_scale, zero_point=W_zp, dtype=torch.qint8)
+
+        # Observe X_fp32 and determine X_scale and X_zero_point, this should match
+        # internals of dynamic linear.
+        X_scale, X_zp = _calculate_dynamic_qparams(X_fp32, torch.quint8)
```

Comment:
Can you add an additional test to check that _calculate_dynamic_qparams works as expected for one sided distributions: please check for both distributions that are positive and distributions that are negative.

Reply:
Sorry I might misunderstand this: we use `_calculate_dynamic_qparams` to calculate the quantization parameters for both activation (`torch.quint8`) and weight (`torch.qint8`). What do "both distributions" refer to here?

As mentioned above, `_calculate_dynamic_qparams` is from https://github.com/pytorch/pytorch/blob/master/torch/quantization/observer.py#L39-L64.

Reply:
For `_calculate_dynamic_qparams`,  we will include 0 in the range. Is this "both distributions" you referred to?

Reply:
Yes, that resolves it.

Reply:
Should I add a test for `_calculate_dynamic_qparams`? The `_calculate_dynamic_qparams` is a test utility function, not a PyTorch API. Let me know if you want me to add a separate test for the test utility function itself.

Reply:
No, there is no need for a test for this function